98_pure_python_eventlog_parsing.md

```diff
@@ -0,0 +1,259 @@
+follow changes in the tooling, and each new version of `tpm2-tools`
+requires us to adjust.
+
+Therefore we propose to replace the approx. 1000 lines of C code in
+the `tpm2-tools` package with a mostly equivalent length (700 lines at
+present) of native Python code, maintained by the Keylime project. The
+code savings come from the fact that OO features of Python allow for
+simpler, more natural code organization; Python `struct` is very
+effective at parsing binary structures into Python native objects; and
+lastly, Python has JSON support.
+
+### Goals
+
+* Improve the stability of measured boot attestation over time.
+* Improve the quality of the measured boot event log parser.
+* Reduce Keylime's dependency on foreign tooling.
+
+### Non-Goals
+
+An explicit non-goal of _this_ proposal is to also enhance the
+`tpm2-tools` package by enumerating the issues found during the
+rewrite. That (entirely reasonable) work item should be handled
+somewhere else.
+
+We do not (in this proposal) aim to change anything else about
+measured boot attestation in Keylime. The python event
+log parser is designed to produce output comparible with
+`tpm2-tools`.
+
+### Notes/Constraints/Caveats (optional)
+
+The most obvious caveat to address is that this project is basically
+duplicating code that already exists in `tpm2-tools`, and therefore
+could be seen as a duplication of effort, unwise both because of the
+wasted labor and the implied long tail of required maintenance.
+
+The developers' argument is that (a) the duplication of effort is not
+that great, since it was done in 2 weeks of part-time work, minus the
+organizational aspects such as writing this document (b) the long tail
+of management is not really that long, because the specifications we
+are following are changing very slowly.
+
+In any case we think that the implicit advantages of python-based
+development (e.g. native JSON support) and the much shorter
+development loop (i.e. the event log parser's developers and the end
+users of Keylime overlap) mitigate the disadvantage of code
+duplication.
+
+### Risks and Mitigations
+
+The largest risk is that the Python based event loop stops being
+maintained, develops bit rot and contributes to a decline of the
+quality of the Keylime code base. Mitigation lays in the decision to
+continue to use JSON as the exchange format between the parser and the
+Keylime measured boot policy agent, which would make a move *back* to
+`tpm2-tools` possible at least in principle.
+
+There are no serious security aspects regarding the event log
+parser. To the extent that the measured boot event log is
+self-checking (i.e. event digests can be recalculated for certain
+types of events), the event log parser is already doing so. NB we are
+not aware of the `tpm2-tools` based event log parser doing any
+self-checking. And of course the python based event log parser is
+capable of generating PCR reference values in the same way
+as `tpm2-tools`.
```

Comment:
The tpm2-tools eventlog parser does validate the entries were possible against the digest and emits an error if it does not match.

Reply:
I have not looked at the `tpm2_eventlog.c` code -- consciously avoided looking there too much except to figure out why generated JSON outputs disagree.

A lot of EFIVAR type entries don't self-validate at the moment, but only when generated by certain BIOSes. No, I don't know why, and I haven't identified the failure parameters yet. Working on making that part of the CI unit tests, while improving the self-checking code that handles {accidentally|deliberately} malformed event logs.